pkg/controllers/user/pipeline/remote/github/github.go

```diff
@@ -0,0 +1,314 @@
+package github
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/remote/model"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"github.com/tomnomnom/linkheader"
+	"golang.org/x/oauth2"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+const (
+	defaultGithubAPI = "https://api.github.com"
+	maxPerPage       = "100"
+	gheAPI           = "/api/v3"
+)
+
+type client struct {
+	Scheme       string
+	Host         string
+	ClientID     string
+	ClientSecret string
+	API          string
+}
+
+func New(pipeline v3.ClusterPipeline) (model.Remote, error) {
+	if pipeline.Spec.GithubConfig == nil {
+		return nil, errors.New("github is not configured")
+	}
+	remote := &client{
+		ClientID:     pipeline.Spec.GithubConfig.ClientID,
+		ClientSecret: pipeline.Spec.GithubConfig.ClientSecret,
+	}
+	if pipeline.Spec.GithubConfig.Host != "" && pipeline.Spec.GithubConfig.Host != "github.com" {
+		remote.Host = pipeline.Spec.GithubConfig.Host
+		if pipeline.Spec.GithubConfig.TLS {
+			remote.Scheme = "https://"
+		} else {
+			remote.Scheme = "http://"
+		}
+		remote.API = remote.Scheme + remote.Host + gheAPI
+	} else {
+		remote.Scheme = "https://"
+		remote.Host = "github.com"
+		remote.API = defaultGithubAPI
+	}
+	return remote, nil
+}
+
+func (c *client) Type() string {
+	return "github"
+}
+
+func (c *client) CanLogin() bool {
+	return true
+}
+
+func (c *client) CanRepos() bool {
+	return true
+}
+func (c *client) CanHook() bool {
+	return true
+}
+
+func (c *client) Login(redirectURL string, code string) (*v3.SourceCodeCredential, error) {
+	githubOauthConfig := &oauth2.Config{
+		RedirectURL:  redirectURL,
+		ClientID:     c.ClientID,
+		ClientSecret: c.ClientSecret,
+		Scopes: []string{"repo",
+			"admin:repo_hook"},
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  fmt.Sprintf("%s%s/login/oauth/authorize", c.Scheme, c.Host),
+			TokenURL: fmt.Sprintf("%s%s/login/oauth/access_token", c.Scheme, c.Host),
+		},
+	}
+
+	token, err := githubOauthConfig.Exchange(oauth2.NoContext, code)
+	if err != nil {
+		logrus.Errorf("Code exchange failed with '%s'\n", err)
```

Comment:
Lets try to avoid logging the error if the error is being passed all the way up to the caller stack. It should be logged there.